### PR TITLE
Poetry

### PR DESCRIPTION
**Pull request by jtpavlock**
## Description

Welcome to my attempt at moving beets towards a more sane dependency management system!

This switches our packaging and dependency management to using [poetry](https://python-poetry.org/). Why the switch? Well for one, poetry works much better as a dependency resolver, and I was able to catch _a lot_ of issues just by importing our current dependencies into it's format and letting it do it's thing. Another thing, is that setuptools has several disadvantages when it comes to alternative package managers. So much so, in fact, that [python officially endorses pipenv](https://packaging.python.org/tutorials/managing-dependencies/) as a setuptools replacement.

> Pipenv is recommended for collaborative projects as it’s a higher-level tool that simplifies dependency management for common use cases.

So, why poetry then? Well, to be honest, I haven't used pipenv before, but I have used poetry and I quite liked it. Also, [a quick google search](https://www.google.com/search?q=pipenv+vs+poetry&oq=pipenv+vs+poetry&aqs=chrome.0.69i59l2j69i61.3464j0j4&sourceid=chrome&ie=UTF-8) will often lead you towards sentiments favoring poetry.

Okay cool, so what does this do for us?

1. As I mentioned above, it is able to better detect dependency conflicts, especially when dealing with various python versions. You'll notice a decent amount of dependency targets have been adjusted to reflect this (turns out 2.7 and 3.4 especially were more broken than we thought). I also adjusted most (all?) of the dependencies to to not automatically upgrade us if a new MAJOR (semver) version of the package was released. Major versions mean breaking changes, so we don't just want to blindly install those.

2. Introduces a `poetry.lock` file. The new flow of creating a development environment is by doing `poetry install -E dev`. This will use the lock file to create a _reproducible_ build inside it's own virtual environment and ensures that all our devs are running the same versions of packages. This can prove useful if you are working on different projects and want to come back to the same environment you were using before. Or if you're debugging it will help eliminate any issues due to package version disparity. There are other uses too I'm sure you'll find well-articulated across the web. You can also run `poetry update` to update any dependencies to the latest supported versions.

3. Removes sphinx as a build-time dependency as per https://discourse.beets.io/t/building-manpages-on-sdist/1443/5 This means tox can now properly build without relying on any external dependencies

4. Removes `setup.py` entirely in place of `pyproject.toml` This also makes us [PEP 518](https://www.python.org/dev/peps/pep-0518/) compliant

5. Tox now installs any extras (normally optional plugin dependencies) that are required for tests to run. No more skipping tests just because the package isn't installed.
    * With this, note that `poetry install -E dev` will only install the _minimum_ required to test and lint normally (aka what we had been doing up until now). Tox will now just install those extra dependencies on its own.

6. There's probably some other stuff I forgot to mention too.

Other things to look into after this:

Ways to automatically bump our dependency targets to new versions as they come available. Maybe something like [dependabot](https://dependabot.com/)?

## To Do

- [ ] Documentation - should just require some changes to contributing.rst
- [ ] Changelog - probably important to note the difference for package managers.. not sure how to best go about this

### CODE REVIEW

**Comment by arogl** [comment]
Like the Black formatter, should we wait until 1.5.0 drops before implementing?

**Comment by jtpavlock** [comment]
I don't actually think black is blocked until 1.5. This one though, perhaps. Especially if it will just add more work for the current release (since I doubt the last thing anyone wants is to delay it any longer). I'm also a little perplexed by the Windows failures, so maybe we will have to wait anyways.

**Comment by arogl** [comment]
Dependency error of pip/setuptools version
```
WARNING: You are using pip version 19.3.1; however, version 20.2 is available  - WINDOWS
ERROR: Package 'setuptools' requires a different Python: 2.7.17 not in '>=3.5'
```

Can we force a pip upgrade as the first collection?
```
Collecting pip
  Downloading pip-20.2-py2.py3-none-any.whl (1.5 MB)    - LINUX
```

File: pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+jellyfish = [
+    { version = '0.6.0', python = '<3.4' },
+    { version = '^0', python = '^3.5' },
+]
+colorama = [
+    { version = '^0', platform = 'win32' },
+]
+
+# optional plugin dependencies
+beautifulsoup4 = { version = '^4', optional = true }
```
[the remark this answers is not in the document]
Comment (jtpavlock) [reply pyproject.toml:56]: So, under `[tool.poetry.dependencies]` is where you specify all version dependencies, optional or not. `[tool.poetry.extras]` is to specify groups of dependencies that you can install with `poetry install -E [extra]` syntax.

So you don't list versions under extras, only names of packages.

**Comment by jtpavlock** [comment]
> * As I [mentioned elsewhere](https://discourse.beets.io/t/building-manpages-on-sdist/1443/4), removing manpages from the sdist will be a disruption to downstream packagers. We should warn them first and give them time to adapt.

I agree. What would be the best way to do this? Do you have a line of contact with the packagers? Or maybe I could make a discourse post accompanied by a changelog entry?

> * One thing to pay attention to here is that the `poetry.lock` style of enforcing consistent dependencies has a downside: most (all?) downstream packagers will not use it. So if there _is_ a problem that's revealed by using an updated version of a dependency, and we only ever run stuff in CI and on developers' machines from the lock file, we will miss those problems before users encounter them. (This is not a problem in the typical use case for stuff like Poetry, where you are developing a server-side app for deployment on your own infrastructure and open-source packaging is not really a concern.) Our current loosey-goosey CI setup has been catching these problems implicitly by just installing what users will get when they `pip install`.

Not necessarily! Because we don't run `poetry install` inside tox, tox will build as normal and will download the most updated packages we specify in our pyproject.toml. The lockfile will only come into play if you run `poetry install` which would be the recommended method for setting up a development environment. You can confirm this behavior by manually editing the lock file to an older dependency and then running tox and comparing the version that tox downloads.

**Comment by arogl** [comment]
> in appveyor.yml add at line 24:
> 
> ```
>     - 'appveyor-retry %PYTHON%/python.exe -m pip install --upgrade pip
> ```

Maybe on a separate line we need to upgrade pip as such
```
     - 'appveyor-retry %PYTHON%/python.exe pip install --upgrade pip
```

I have found some strange failures when trying to upgrade pip via the python -m pip  command.

**Comment by jtpavlock** [comment]
@arogl it doesn't seem to want to cooperate

**Comment by arogl** [comment]
> 
> 
> @arogl it doesn't seem to want to cooperate. In the end (assuming 1.5 is near), I think this is better suited to be after that release, if not only to give the package managers time to adjust to the new system.

I can see that and I am running out of ideas. 

Currently on my work machine so I cannot test.

**Comment by jtpavlock** [comment]
> How do we see the older logs?

If you click on the ❌ next to the individual commits you can see the CI builds for them

**Comment by arogl** [comment]
> 
> 
> > How do we see the older logs?
> 
> If you click on the ❌ next to the individual commits you can see the CI builds for them

Got it.

Can we try this one above the current tox line?

```
  - 'appveyor-retry %PYTHON%/python.exe -m pip install --upgrade pip
```

**Comment by arogl** [comment]
Looks like we are back to the setuptools error 👎

It seems we might need to limit setuptools with ==30.0?

Trying to find a version of setuptools that runs  with 2.7, I am not sure what version is installed in the linux version.

**Comment by sampsyo** [comment]
> Do you have a line of contact with the packagers? Or maybe I could make a discourse post accompanied by a changelog entry?

I don't anymore, unfortunately. Starting with a Discourse post would be great, but I think it might also be polite to actually track down the maintainers for major packaging systems and shoot them an email too, inviting them to weigh in on the Discourse thread or an issue.

> Not necessarily for tox! Because we don't run `poetry install` inside tox, tox will build as normal and will download the most updated packages we specify in our pyproject.toml. The lockfile will only come into play if you run `poetry install` which would be the recommended method for setting up a development environment.

That's a good point; tox and CI could be set up to do the "normal" thing. But I guess it seems a little less clear to me then what the value is of having the recommended development environment be based on the lock file… if the aim is to stick to "known good" versions of dependencies when later versions might cause breakage, isn't that sort of a negative rather than a positive?

I think the lock file approach for development makes a lot of sense when the ultimate deployment destination will _also_ use a lock file. But it seems like it loses some of its appeal when the ultimate deployment is on users' machines that _won't_ be using our lock file. I know it's a super popular approach (Yarn, npm, Cargo, etc. all use it now!) so I wonder if there isn't some good writing out there we could read about this issue: when/whether it makes sense to check the lock file into git when you're building a user-facing app for open-source distribution.

> Eventually, I'd like decouple our cli application and the api/library logic into separate repos, this way we could have pinned dependencies for the cli, in combination with something like [pipx](https://github.com/pipxproject/pipx) so that users could also match the dependency versions.

Further decoupling sounds great, certainly! Although I would argue that separating the actual git repositories is not necessary to achieve such a decoupling, FWIW.

**Comment by arogl** [comment]
> 
> 
> Looks like we are back to the setuptools error 👎
> 
> It seems we might need to limit setuptools with <=42.0.2 on python 2.7 on windows?
> 
> ```
>   - 'appveyor-retry %PYTHON%/python.exe -m pip install --upgrade setuptools<=42.0.2
> ```

On windows it seems to do another install of setuptools somewhere from v41..2.0 to v45.0.0 somewhere.

**Comment by jtpavlock** [comment]
> I think it might also be polite to actually track down the maintainers for major packaging systems and shoot them an email too

Good idea, I'll see if I can track them down before implementing this. Speaking of, would you like to wait until after we release 1.5.0 to merge this?

I tend to agree with you about the lock file, so I took it back out and added it to the gitignore https://github.com/beetbox/beets/pull/3715/commits/2fe2d88d9ad10e5227dbf3bc09a750c157065ba0. If someone else comes along with a more convincing argument for committing it to version control, then we can reassess.

**Comment by arogl** [comment]
In the Appveyor logs for py2.7 the time references indicate in the 13 seconds setuptools seems to be updated from 41.2.0 to 45.0.0. I cannot see in the log where that has occurred
```
[00:01:04]   Downloading pluggy-0.13.1-py2.py3-none-any.whl (18 kB)
[00:01:04] Requirement already satisfied: virtualenv>=1.11.2 in c:\python27\lib\site-packages (from tox<=3.8.1) (15.0.1)
[00:01:04] Requirement already satisfied: setuptools>=30.0.0 in c:\python27\lib\site-packages (from tox<=3.8.1) (41.2.0)
```
```
[00:01:17]   Downloading attrs-19.3.0-py2.py3-none-any.whl (39 kB)
[00:01:17] Requirement already satisfied: setuptools in c:\projects\beets\.tox\.package\lib\site-packages (from jsonschema<4.0,>=3.1->poetry>=0.12) (45.0.0)
[00:01:17] ERROR: Package 'setuptools' requires a different Python: 2.7.17 not in '>=3.5'
```

**Comment by arogl** [comment]
Setting up a python 2.7.18 windows environment to test

**Pull request closed without merging**